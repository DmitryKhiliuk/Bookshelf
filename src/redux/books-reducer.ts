import {v1} from "uuid";

export type BookType = {
    id: string
    title: string
    author: string
    year: string
    image: string
}

export type ActionType = addBookACType | removeBookACType | editBookACType

const initialState: BookType[] = [
    {id:v1(), title: 'React: современные шаблоны для разработки ', author: 'Алекс Бэнкс, Ева Порселло', year: '2021', image: 'https://s4-goods.ozstatic.by/480/537/52/101/101052537_0_React_sovremennie_shabloni_dlya_razrabotki_prilozheniy_Aleks_Benks_Eva_Porsello.jpg'},
    {id:v1(), title: 'JavaScript с нуля', author: 'Кирупа Чиннатамби', year: '2021', image: 'https://s2-goods.ozstatic.by/480/443/28/101/101028443_0_JavaScript_s_nulya_Kirupa_CHinnatambi.jpg'},
    {id:v1(), title: 'Turbo Pascal', author: 'Валерий Фаронов', year: '2021', image: 'https://s3-goods.ozstatic.by/480/884/110/101/101110884_0_Turbo_Pascal_Uchebnoe_posobie_Valeriy_Faronov.jpg'},
    {id:v1(), title: 'Биполярное расстройство', author: 'Е. Касьянов', year: '2022', image: 'https://s2-goods.ozstatic.by/480/803/109/101/101109803_0_Bipolyarnoe_rasstroystvo_gid_po_vizhivaniyu_dlya_teh_kto_chasto_ne_vidit_beloy_polosi_Masha_Pushkina_E_Kasyanov.jpg'},
]

export const booksReducer = (state:BookType[] = initialState, action: ActionType):BookType[] => {
    switch (action.type) {
        case "ADD-BOOK": {
            return [action.newBook, ...state]
        }
        case "REMOVE-BOOK": {
            return state.filter((b) => b.id !== action.id)
        }
        case "EDIT-BOOK": {
            return state.map((b) => b.id === action.id ? {...b, ...action.newBook} : b)
        }
    }
    return state
}

export type addBookACType = ReturnType<typeof addBookAC>
export const addBookAC = (newBook:BookType) => {
    return {
        type: 'ADD-BOOK',
        newBook
    } as const
}
export type removeBookACType = ReturnType<typeof removeBookAC>
export const removeBookAC = (id:string) => {
    return {
        type: 'REMOVE-BOOK',
        id
    } as const
}
export type editBookACType = ReturnType<typeof editBookAC>
export const editBookAC = (id:string, newBook:BookType) => {
    return {
        type: 'EDIT-BOOK',
        id,
        newBook
    } as const
}