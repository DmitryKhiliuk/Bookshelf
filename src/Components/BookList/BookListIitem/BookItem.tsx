import React from 'react';
import {BookType, removeBookAC} from "../../../redux/books-reducer";
import {Button} from "../../Button";
import s from "./BookListItem.module.css"
import {NavLink} from "react-router-dom";
import {useDispatch} from "react-redux";


export type BookItemType = {
    book: BookType
}

export const BookItem = (props: BookItemType) => {

    const dispatch = useDispatch()

    const onClickHandler = () => {
        dispatch(removeBookAC(props.book.id))
    }

    return (
        <div className={s.bookItem}>
            <div><img src={props.book.image} alt="img"/></div>
            <div>
                <div>{props.book.title}</div>
                <div>{props.book.author}</div>
                <div>{props.book.year}</div>
            </div>
            <div>
                <div><NavLink to={'/edit-book'}><Button callBack={onClickHandler} name={'Редактировать'}/></NavLink></div>
                <Button callBack={onClickHandler} name={'Удалить'}/>
            </div>
        </div>
    );
};

