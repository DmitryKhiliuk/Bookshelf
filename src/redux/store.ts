import {combineReducers, createStore} from "redux";
import {booksReducer} from "./books-reducer";
import {loadState, saveState} from "./localstorage-utils";

export const rootReducer = combineReducers({
    books: booksReducer
})

export const store = createStore(rootReducer, loadState())

store.subscribe(() => {
    saveState({
        books:store.getState().books
    })
})

export type AppRootStateType = ReturnType<typeof rootReducer>

// @ts-ignore
window.store = store;