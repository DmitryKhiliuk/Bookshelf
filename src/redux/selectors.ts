import {AppRootStateType} from "./store";

export const selectBooks = (state: AppRootStateType) => state.books