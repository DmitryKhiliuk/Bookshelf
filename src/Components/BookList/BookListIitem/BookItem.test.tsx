import React from "react";
import {BookItem} from "./BookItem";
import {render, screen} from "@testing-library/react";
import {BookType} from "../../../redux/books-reducer";
import {MemoryRouter} from "react-router-dom";
import * as reduxHooks from "react-redux";
import userEvent from "@testing-library/user-event";


jest.mock('react-redux')
const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch')
const dispatch = jest.fn()

jest.mock('console')
const mockedLog = jest.spyOn(console, 'log')

let bookItem: BookType

describe('book item test', () => {
    beforeEach(() => {
        bookItem = {
            id: '1',
            title: "Book",
            author: "writer",
            year: "1900",
            image: "url",
        }
    })
    it('should create book item', () => {
        const view = render(
            <MemoryRouter>
                <BookItem book={bookItem}/>
            </MemoryRouter>
        )
        expect(view).toMatchSnapshot()
    })
    it('should render buttons', () => {
        render(<MemoryRouter><BookItem book={bookItem}/></MemoryRouter>)
        const deleteBtn = screen.getByText('Удалить')
        const editBtn = screen.getByText('Редактировать')
        const allButton = screen.getAllByRole('button')
        expect(allButton).toEqual([editBtn,deleteBtn])
    })
    it('should dispatch remove book action', () => {
        mockedDispatch.mockReturnValue(dispatch)
        render(<MemoryRouter><BookItem book={bookItem}/></MemoryRouter>)
        userEvent.click(screen.getByText('Удалить'))
        expect(dispatch).toHaveBeenCalledTimes(1)
    })
    it('main page button test', () => {
        mockedLog.mockReturnValue()
        render(<MemoryRouter><BookItem book={bookItem}/></MemoryRouter>)
        userEvent.click(screen.getByText('Редактировать'))
        expect(mockedLog).toHaveBeenCalled()
    })

})