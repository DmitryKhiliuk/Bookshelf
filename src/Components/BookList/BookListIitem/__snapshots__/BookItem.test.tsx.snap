// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book item test  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bookItem"
      >
        <div>
          <img
            alt="img"
            src="url"
          />
        </div>
        <div
          class="bookItem__text"
        >
          <div
            class="title"
          >
            Book
          </div>
          <div
            class="author"
          >
            writer
          </div>
          <div
            class="year"
          >
            1900
          </div>
        </div>
        <div
          class="bookItem__button"
        >
          <div>
            <a
              class=""
              href="/edit-book/1"
            >
              <div>
                <button
                  class="button"
                >
                  Редактировать
                </button>
              </div>
            </a>
          </div>
          <div
            class="bookItem__button__delete"
          >
            <div>
              <button
                class="button"
              >
                Удалить
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bookItem"
    >
      <div>
        <img
          alt="img"
          src="url"
        />
      </div>
      <div
        class="bookItem__text"
      >
        <div
          class="title"
        >
          Book
        </div>
        <div
          class="author"
        >
          writer
        </div>
        <div
          class="year"
        >
          1900
        </div>
      </div>
      <div
        class="bookItem__button"
      >
        <div>
          <a
            class=""
            href="/edit-book/1"
          >
            <div>
              <button
                class="button"
              >
                Редактировать
              </button>
            </div>
          </a>
        </div>
        <div
          class="bookItem__button__delete"
        >
          <div>
            <button
              class="button"
            >
              Удалить
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`book item test should create book item 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bookItem"
      >
        <div>
          <img
            alt="img"
            src="url"
          />
        </div>
        <div
          class="bookItem__text"
        >
          <div
            class="title"
          >
            Book
          </div>
          <div
            class="author"
          >
            writer
          </div>
          <div
            class="year"
          >
            1900
          </div>
        </div>
        <div
          class="bookItem__button"
        >
          <div>
            <a
              class=""
              href="/edit-book/1"
            >
              <div>
                <button
                  class="button"
                >
                  Редактировать
                </button>
              </div>
            </a>
          </div>
          <div
            class="bookItem__button__delete"
          >
            <div>
              <button
                class="button"
              >
                Удалить
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bookItem"
    >
      <div>
        <img
          alt="img"
          src="url"
        />
      </div>
      <div
        class="bookItem__text"
      >
        <div
          class="title"
        >
          Book
        </div>
        <div
          class="author"
        >
          writer
        </div>
        <div
          class="year"
        >
          1900
        </div>
      </div>
      <div
        class="bookItem__button"
      >
        <div>
          <a
            class=""
            href="/edit-book/1"
          >
            <div>
              <button
                class="button"
              >
                Редактировать
              </button>
            </div>
          </a>
        </div>
        <div
          class="bookItem__button__delete"
        >
          <div>
            <button
              class="button"
            >
              Удалить
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
